cmd/tailscale/cli/cli.go

```diff
@@ -216,11 +241,68 @@ var rootArgs struct {
 	socket string
 }
 
-func walkCommands(cmd *ffcli.Command, f func(*ffcli.Command)) {
-	f(cmd)
-	for _, sub := range cmd.Subcommands {
-		walkCommands(sub, f)
+type cmdWalk struct {
+	cmd     *ffcli.Command
+	parents []*ffcli.Command
+}
+
+func (w cmdWalk) Path() string {
+	if len(w.parents) == 0 {
+		return w.cmd.Name
+	}
+
+	var sb strings.Builder
+	for _, p := range w.parents {
+		sb.WriteString(p.Name)
+		sb.WriteString(" ")
+	}
+	sb.WriteString(w.cmd.Name)
+	return sb.String()
+}
+
+func walkCommands(cmd *ffcli.Command, f func(w cmdWalk) bool) {
+	var i int
+	subs := [][]*ffcli.Command{{cmd}}
+	path := []*ffcli.Command{nil}
+	for len(subs) > 0 {
+		// Trim empty subs.
+		for len(subs[i]) == 0 {
+			if i == 0 {
+				return
+			}
+			i--
+		}
+		subs = subs[:i+1]
+		path = path[:i]
+
+		// Pop the next cmd.
+		cmd := subs[i][0]
+		subs[i] = subs[i][1:]
+		if !f(cmdWalk{cmd, path}) {
+			return
+		}
```

Comment:
the recursion was kinda nice in comparison? 😅

Reply:
So much nicer 😆